pmlc/dialect/pxa/transforms/reorder_layout.cc

```diff
@@ -0,0 +1,264 @@
+// Copyright 2020 Intel Corporation
+
+#include <numeric>
+#include <vector>
+
+#include "mlir/Dialect/Affine/IR/AffineMemoryOpInterfaces.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/IR/IntegerSet.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/DebugStringHelper.h"
+
+#include "pmlc/dialect/pxa/transforms/pass_detail.h"
+#include "pmlc/dialect/pxa/transforms/passes.h"
+#include "pmlc/util/logging.h"
+
+namespace pmlc::dialect::pxa {
+
+namespace {
+
+class ReorderLayoutPass final : public ReorderLayoutBase<ReorderLayoutPass> {
+public:
+  void runOnFunction();
+};
+
+struct ReorderDesc {
+  mlir::AffineMap reorderMap;
+  mlir::SmallVector<uint64_t, 6> reorderedShape;
+  mlir::SmallVector<uint64_t, 6> reorderedVector;
+};
+
+/// Simplify affine map given integral constraints. Returns the same map if it
+/// cannot be simplified further.
+///
+/// Input:
+///   map         = (d0, d1) -> (d0 + d1, (d0 * 16 + d1 * 8) floordiv 8 floordiv
+///   2,
+///                              (d0 * 16 + d1 * 8) floordiv 8 % 2, 0)
+///   constraints = {0 <= d1 < 2, 0 <= d0 < 6}
+///
+/// Output:
+///   (d0, d1) -> (d0 + d1, d0, d1, 0)
+mlir::AffineMap simplifyConstrainedAffineMap(mlir::AffineMap map,
+                                             mlir::IntegerSet constraints) {
+  // TODO: Implement.
+  return map;
```

Comment:
What is the strategy for implementation on this?  A set of rewrites or something more general (such as flattening to a system of pure affines with slack variables or the like)?  Have you examined the MLIR affine code to see if there is existing code for any of this?  Looking at the above, I can see how this might be done via rewrites, but it seems potentially brittle.

Reply:
I was thinking about rewrites with matching some patterns, as that seems to be enough for all maps I saw generated in our code. MLIR already has simplification without constraints via flattening, so that could be used as a cleanup after our rewrites.
I looked briefly at what is directly exposed in MLIR and there doesn't seem to be anything like this. Some implementation details used in MLIR look useful though, so I think it may be good to upstream such simplification in future.